cli/src/program.rs

```diff
@@ -1129,22 +1289,173 @@ fn process_program_deploy(
+        )],
+        Some(&fee_payer_signer.pubkey()),
+        &blockhash,
+    );
+
+    if sign_only {
+        let mut tx = Transaction::new_unsigned(final_message);
+        let signers = &[fee_payer_signer, upgrade_authority_signer];
+        // Using try_partial_sign here because fee_payer_signer might not be the fee payer we
+        // end up using for this transaction (it might be NullSigner).
+        tx.try_partial_sign(signers, blockhash)?;
+        return_signers_with_config(
+            &tx,
+            &config.output_format,
+            &ReturnSignersConfig {
+                dump_transaction_message: true, // print out resulting signature
+            },
+        )
+    } else {
+        let final_message = Some(final_message);
+
+        let buffer_len = {
+            // Check supplied buffer account.
+            if let Some(account) = rpc_client
+                .get_account_with_commitment(&buffer_pubkey, config.commitment)?
+                .value
+            {
+                if !bpf_loader_upgradeable::check_id(&account.owner) {
+                    return Err(format!(
+                        "Buffer account {buffer_pubkey} is not owned by the BPF Upgradeable Loader",
+                    )
+                    .into());
+                }
+
+                match account.state() {
+                    Ok(UpgradeableLoaderState::Buffer { .. }) => {
+                        // continue if buffer is initialized
+                    }
+                    Ok(UpgradeableLoaderState::Program { .. }) => {
+                        return Err(format!(
+                            "Cannot use program account {buffer_pubkey} as buffer"
+                        )
+                        .into());
+                    }
+                    Ok(UpgradeableLoaderState::ProgramData { .. }) => {
+                        return Err(format!(
+                            "Cannot use program data account {buffer_pubkey} as buffer",
+                        )
+                        .into())
+                    }
+                    Ok(UpgradeableLoaderState::Uninitialized) => {
+                        return Err(
+                            format!("Buffer account {buffer_pubkey} is not initialized").into()
+                        );
+                    }
+                    Err(_) => {
+                        return Err(format!(
+                            "Buffer account {buffer_pubkey} could not be deserialized"
+                        )
+                        .into())
+                    }
+                };
+
+                account
+                    .data
+                    .len()
+                    .saturating_sub(UpgradeableLoaderState::size_of_buffer_metadata())
+            } else {
+                return Err(format!(
+                    "Buffer account {buffer_pubkey} not found, was it already consumed?",
+                )
+                .into());
+            }
+        };
+
```

Comment:
As [per your comment below](https://github.com/solana-labs/solana/pull/33860#discussion_r1412538579), we don't need this code in `process_program_upgrade` - hence I'm removing it,

so there is no longer a case for a "reusable function", and also note that `process_program_deploy` validates buffer account differently from how `process_write_buffer` validates it for example, meaning in general each needs it's own validation function,

let me know you still want this code moved into separate function(s)

Reply:
This function is still useful -- it's not only returning the size of the buffer, but also checking to make sure that the provided buffer pubkey is actually a buffer account.

Since we want to catch if people use an incorrect for the buffer and give them a nice error, I'm still in favor of refactoring since it into a separate function

Reply:
Probably makes sense to do this refactoring in a separate PR (since it's not even need for this PR now): https://github.com/solana-labs/solana/pull/34317 ? If that's what you had in mind.